src/dataflow/src/source/kafka.rs

```diff
@@ -186,25 +137,105 @@ impl KafkaSourceReader {
+        let partition_info = Arc::new(Mutex::new(None));
+        {
+            let partition_info = Arc::downgrade(&partition_info);
+            let topic = topic.clone();
+            let consumer = Arc::clone(&consumer);
+            let metadata_refresh_frequency = config_options
+                .get("topic.metadata.refresh.interval.ms")
+                // Safe conversion: statement::extract_config enforces that option is a value
+                // between 0 and 3600000
+                .map(|s| Duration::from_millis(s.parse().unwrap()))
+                // Default value obtained from https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md
+                .unwrap_or_else(|| Duration::from_secs(300));
+
+            std::thread::Builder::new()
+                .name("kafka-metadata".to_string())
+                .spawn(move || {
+                    while let Some(partition_info) = partition_info.upgrade() {
+                        let info = get_kafka_partitions(&consumer, &topic, Duration::from_secs(30));
+                        *partition_info.lock().unwrap() = Some(info);
+                        std::thread::sleep(metadata_refresh_frequency);
+                    }
+                })
+                .unwrap();
+        }
 
         KafkaSourceReader {
             topic_name: topic,
             source_name,
             id: source_id,
             partition_consumers: VecDeque::new(),
-            known_partitions: 0,
-            consumer: Arc::new(consumer),
+            consumer,
             worker_id,
+            worker_count,
             last_offsets: HashMap::new(),
             start_offsets,
             logger,
             stats_rx,
             last_stats: None,
+            partition_info,
         }
     }
 
+    /// Ensures that a partition queue for `pid` exists.
+    /// In Kafka, partitions are assigned contiguously. This function consequently
+    /// creates partition queues for every p <= pid
+    fn add_partition(&mut self, pid: PartitionId) {
+        if !crate::source::responsible_for(
+            &self.id.source_id,
+            self.worker_id,
+            self.worker_count,
+            &pid,
+        ) {
+            return;
+        }
+        let pid = match pid {
+            PartitionId::Kafka(p) => p,
+            _ => unreachable!(),
+        };
+        if self.last_offsets.contains_key(&pid) {
+            return;
+        }
+
+        // Indicate a last offset of -1 if we have not been instructed to have a specific start
+        // offset for this topic.
+        let start_offset = match self.start_offsets.get(&pid) {
+            // Seek to the *next* offset (aka start_offset + 1) that we have not yet processed
+            Some(offset) => *offset + 1,
+            None => 0,
+        };
+
+        self.create_partition_queue(pid, Offset::Offset(start_offset));
+
+        let prev = self.last_offsets.insert(pid, start_offset - 1);
```

Comment:
why do we use `offset - 1` and `offset` in different places instead of converging around one convention?